gto-support-kotlin-coroutines/src/test/java/org/ccci/gto/android/common/kotlin/coroutines/ReadWriteMutexTest.kt

```diff
@@ -0,0 +1,114 @@
+package org.ccci.gto.android.common.kotlin.coroutines
+
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.yield
+import org.junit.Test
+
+class ReadWriteMutexTest {
+    private val mutex = ReadWriteMutex()
+
+    @Test
+    fun testWriteExclusion() {
+        runBlocking {
+            launch {
+                expect(3)
+                mutex.write.withLock {
+                    expect(5)
+                }
+            }
+
+            expect(1)
+            mutex.write.withLock {
+                expect(2)
+                yield()
+                expect(4)
+            }
+        }
+        finish(6)
+    }
+
+    @Test
+    fun testReadShared() {
+        runBlocking {
+            launch {
+                mutex.read.withLock {
+                    yield()
+                    expect(2)
+                    yield()
+                    expect(4)
+                }
+            }
+
+            launch {
+                mutex.read.withLock {
+                    expect(1)
+                    yield()
+                    expect(3)
+                }
+            }
```

Comment:
This seems to assume that the first `launch` will run first, and yield to the second one. Is that a safe assumption?
I don't really know the scheduling rules for coroutines.

I noticed if I put `delay(1L)` right before the first `launch`, the test fails. But if I change it to `delay(0L)`, it passes.

Reply:
within `runBlocking` coroutines are launched on the blocking thread by default. so because I'm not specifying an alternate dispatcher for any of the `launch` calls, the coroutines will be queued up in order. using `yield()` will bump the current coroutine to the end of the queue.

I think the `delay(0L)` behaves that way because I call `yield()` at line 39, whereas `delay(1L)` has to delay the coroutine by at least 1 ms, so the `yield()` at line 39 has nothing to actually yield to and execution continues.

Reply:
I can probably clean this test up a bit to make it more straightforward in what it's testing.

Reply:
from the `runBlocking` documentations:
>  The default [CoroutineDispatcher] for this builder is an internal implementation of event loop that processes continuations in this blocked thread until the completion of this coroutine.

Reply:
I liked the way you cleaned this up; I think that addresses my concern here.

"event loop" must be the promise you can rely on, that these things are done in order, on this thread.

I think that also must mean that the 'AtomicInteger' could actually be a normal int, right? There's not actually any multi-threadedness going on in this test.
(Though maybe makes sense to keep AtomicInteger because it's a little nicer to read.)

Reply:
that is true it could be a regular int and a regular boolean. I had copied the `expect()` and `finish()` methods from [here](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/jvm/test/TestBase.kt#L97-L119).